### ISSUE TITLE
Edge case on special `array` for `lengths_linter()`

### ISSUE THREAD

**Issue by etiennebacher** [r-lib/lintr#2649]
Found this edge case quite deep in a package and it's hard to come up with less convoluted example. 

`lintr` suggests using `lengths()`:

``` r
  lintr::lint(
    text = '
test <- array(
  list(NULL, c("max1", "min1")),
  dim = 2L
) |>
  structure(dimnames = list(c(NA, "data1")))

map_int(test, length)
  ',
    linters = lintr::lengths_linter()
  )
#> <text>:8:1: warning: [lengths_linter] Use lengths() to find the length of each element in a list.
#> map_int(test, length)
#> ^~~~~~~~~~~~~~~~~~~~~
```

But it would give an object with different attributes:

``` r
test <- array(
  list(NULL, c("max1", "min1")),
  dim = 2L
) |>
  structure(dimnames = list(c(NA, "data1")))

waldo::compare(
  purrr::map_int(test, length),
  lengths(test)
)
#> `names(old)` is a character vector (NA, 'data1')
#> `names(new)` is absent
#> 
#> `dim(old)` is absent
#> `dim(new)` is an integer vector (2)
#> 
#> `dimnames(old)` is absent
#> `dimnames(new)` is a list
```

**Comment by MichaelChirico** [comment]
I would just mark this as `# nolint`. We might put a note in the documentation... array-of-lists is a strange object. Whether to retain `dim()` in this case seems unclear.

**Comment by MichaelChirico** [comment]
This may at root be an R bug. I filed this report:

https://bugs.r-project.org/show_bug.cgi?id=18858

If it's intended behavior, here is a proposed patch for our docs:

```diff
diff --git a/R/lengths_linter.R b/R/lengths_linter.R
index 3ac9428a..67d73deb 100644
--- a/R/lengths_linter.R
+++ b/R/lengths_linter.R
@@ -1,8 +1,13 @@
 #' Require usage of `lengths()` where possible
 #'
 #' [lengths()] is a function that was added to base R in version 3.2.0 to
-#'   get the length of each element of a list. It is equivalent to
-#'   `sapply(x, length)`, but faster and more readable.
+#'   get the length of each element of a list. It is almost always equivalent
+#'   to `sapply(x, length)`, but faster and more readable.
+#'
+#' One known case where this does not strictly apply is to an array of lists,
+#'   e.g. `array(list(1, 1, 1, 1:2), dim = c(2L, 2L))`, where `lengths()`
+#'   will retain the shape of its input where `sapply()` will not. Whether
+#'   that's a good thing may be context-dependent.
 #'
 #' @examples
 #' # will produce lints
diff --git a/man/lengths_linter.Rd b/man/lengths_linter.Rd
index 82a61cb0..b6710dfc 100644
--- a/man/lengths_linter.Rd
+++ b/man/lengths_linter.Rd
@@ -8,8 +8,14 @@ lengths_linter()
 }
 \description{
 \code{\link[=lengths]{lengths()}} is a function that was added to base R in version 3.2.0 to
-get the length of each element of a list. It is equivalent to
-\code{sapply(x, length)}, but faster and more readable.
+get the length of each element of a list. It is almost always equivalent
+to \code{sapply(x, length)}, but faster and more readable.
+}
+\details{
+One known case where this does not strictly apply is to an array of lists,
+e.g. \code{array(list(1, 1, 1, 1:2), dim = c(2L, 2L))}, where \code{lengths()}
+will retain the shape of its input where \code{sapply()} will not. Whether
+that's a good thing may be context-dependent.
 }
 \examples{
 # will produce lints
```
